src/compiler/checker.ts

```diff
@@ -12941,7 +12939,15 @@ namespace ts {
                         continue;
                     }
                 }
-                result.push(getSignatureFromDeclaration(decl));
+                // If this is a function or method declaration, get the accurate minArgumentCount from the type tag, if present.
+                // If this is a variable or property declaration, apply the type tag to the target
+                // (getTypeForVariableLikeDeclaration()), not to the initializer.
+                result.push(
+                    (!isFunctionExpressionOrArrowFunction(decl) &&
+                        !isObjectLiteralMethod(decl) &&

```

Comment:
How are object literal methods different from normal methods? Shouldn't both allow you to specify `@type` in the same way as a function declaration?

On that note, I think we need a test of class methods.

Reply:
I'm not 100% on the difference between object literal methods and normal methods --- I speculate it has something to do with, normal methods can have multiple, merged declarations (like functions) and object literal methods can't (like variables)?
```ts repro
// @errors: 2300

declare class Normal {
  method(arg: "b"): void;
}
interface Normal {
  method(arg: "a"): void;
}

// error TS2300: Duplicate identifier 'Literal'.
type Literal = {
  method(arg: "b"): void;
};
// error TS2300: Duplicate identifier 'Literal'.
interface Literal {
  method(arg: "a"): void;
}
```
[Workbench repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/PTAEAEFMCdoe2gZwFygEwGYAMWBQuATSAYwBsBDaSUM8xRUAOQQFtzTQBvXUUFyAC4ALOAQAUlAOaoARACMZASlQA3OAEsCAblwBfXOoB2AmADNyxas2hsO3Xv2GiJ0aaBnklqjdr34QoDDw0KAAKgDKmDioACIArgAOpOrE5CagmpDG6qbqMKAA5AAy6ibQ7AUAdLgCAJ4J1CVl7KAAvFw8fIIi4lKyCsqgapo6ujoBQQhhkdhYsYnJqemZ2bn5xaUwFdVGZeaWoE1bdp2OPS5uHl5DPqNAA).

What I *have* confirmed is:
- Object literal methods:
  - Already obey `@type` tags, in getTypeOfVariableOrParameterOrPropertyWorker():
    ```ts
    else if (isObjectLiteralMethod(declaration)) {
      type = tryGetTypeFromEffectiveTypeNode(declaration) || checkObjectLiteralMethod(declaration, CheckMode.Normal);
    }
    ```
  - Error if an initializer isn't compatible with its target.
- Normal methods:
  - Ignore `@type` tags.
  - Error if an initializer isn't compatible with its target, but via specialized logic in checkFunctionOrMethodDeclaration():
    ```ts
    if (typeTag && typeTag.typeExpression && !getContextualCallSignature(getTypeFromTypeNode(typeTag.typeExpression), node)) {
      error(typeTag.typeExpression.type, Diagnostics.The_type_of_a_function_declaration_must_match_the_function_s_signature);
    }
    ```
    Object literal methods call checkObjectLiteralMethod() instead.

By exempting variables, properties, and object literal methods this PR:
- Makes functions and normal methods obey `@type` tag, like variables, properties, and object literal methods already do.
- Variables, properties, and object literal methods continue to error if an initializer isn't compatible with its target.

> On that note, I think we need a test of class methods.

Is the existing one sufficient? https://github.com/microsoft/TypeScript/blob/main/tests/cases/conformance/jsdoc/checkJsdocTypeTag7.ts

Reply:
No, it doesn't test optionality. I manually tested it and it doesn't work.

Reply:
- [x] I've now added a test for class methods and optional parameters.